FreeRTOS/FreeRTOS-Plus-TCP
# Reparations and recommendations for the dev/IPv6_integration branch

Description
-----------
Nost changes in this branch are related to DNS.

- Renamed pCallback to pCallbackFucntion for clarity

- Allows for multiple DNS server, hopping to the next one after a failure.

- Added FreeRTOS_getaddrinfo() and FreeRTOS_getaddrinfo_a() to get multiple find results and also to return IPv6 addresses.

- Started using the new unified 'struct freertos_sockaddr', which is downward compatible with the labs/IPv6_multi branch.

- A new function prvPrepare_ReadIPAddress() reads a string and see if it is a valid IP-address.

- Here is the new flow of functions in DNS:

~~~
FreeRTOS_getaddrinfo()
    prvPrepareLookup()
        prvPrepare_ReadIPAddress() ? resolve "192.168.2.1" to an IP-address
        Prepare_CacheLookup() ? try to find the host name in DNS cache
        If not resolve, call prvGetHostByName()
~~~	
- Added FreeRTOS_IPStart() and kept the old 'FreeRTOS_IPInit()' for backward compatibility in case 'ipconfigCOMPATIBLE_WITH_SINGLE' is defined.

- I tried to get rid of old macros `ipLOCAL_IP_ADDRESS_POINTER` and `ipLOCAL_MAC_ADDRESS` and replace them with references to end-points. Job still unfinished.

~~~diff
-    pvCopySource = ipLOCAL_MAC_ADDRESS;
+    pvCopySource = pxTargetEndPoint->xMACAddress.ucBytes;
~~~

Also I replaced these calls:

~~~diff
-    ( void ) xNetworkInterfaceOutput( pxNetworkBuffer, pdTRUE );
+    pxEndPoint->pxNetworkInterface->pfOutput( pxEndPoint->pxNetworkInterface, pxNetworkBuffer, pdTRUE );
~~~

Places some #idef's at the beginning of a module like e.g. :
~~~
/* *INDENT-OFF* */
#if( ipconfigUSE_IPv6 != 0 )
/* *INDENT-ON* */
    ...
/* *INDENT-OFF* */
#endif /* ipconfigUSE_IPv6 != 0 */
/* *INDENT-ON* */
~~~


Test Steps
-----------
I used the [IPv6_Multi_WinSim_demo demo](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/tree/labs/ipv6_multi/demos/IPv6_Multi_WinSim_demo), but it might need some more adaptations.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
